File: templates/problems/09_heap/07_find_median_stream/find_median_stream_test.go
package find_median_stream

import (
	"math"
	"testing"
)

func TestMedianFinder(t *testing.T) {
	tests := []struct {
		name    string
		adds    []int
		medians []float64
	}{
		{
			name:    "example from problem",
			adds:    []int{1, 2, 3},
			medians: []float64{1.0, 1.5, 2.0},
		},
		{
			name:    "single element",
			adds:    []int{5},
			medians: []float64{5.0},
		},
		{
			name:    "two elements",
			adds:    []int{1, 2},
			medians: []float64{1.0, 1.5},
		},
		{
			name:    "negative numbers",
			adds:    []int{-1, -2, -3},
			medians: []float64{-1.0, -1.5, -2.0},
		},
		{
			name:    "duplicates",
			adds:    []int{1, 1, 1, 1},
			medians: []float64{1.0, 1.0, 1.0, 1.0},
		},
		{
			name:    "descending order",
			adds:    []int{5, 4, 3, 2, 1},
			medians: []float64{5.0, 4.5, 4.0, 3.5, 3.0},
		},
		{
			name:    "large gap",
			adds:    []int{0, 100},
			medians: []float64{0.0, 50.0},
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			mf := ConstructorMedianFinder()
			for i, num := range tt.adds {
				mf.AddNum(num)
				got := mf.FindMedian()
				if math.Abs(got-tt.medians[i]) > 1e-5 {
					t.Errorf("after AddNum(%d), FindMedian() = %f, want %f", num, got, tt.medians[i])
				}
			}
		})
	}
}

File: templates/problems/09_heap/07_find_median_stream/typescript/find_median_stream.test.ts
import { describe, it, expect } from "vitest";
import { MedianFinder } from "./find_median_stream";

describe("MedianFinder", () => {
  it("example from problem", () => {
    const mf = new MedianFinder();
    mf.addNum(1);
    expect(mf.findMedian()).toBe(1.0);
    mf.addNum(2);
    expect(mf.findMedian()).toBe(1.5);
    mf.addNum(3);
    expect(mf.findMedian()).toBe(2.0);
  });

  it("single element", () => {
    const mf = new MedianFinder();
    mf.addNum(5);
    expect(mf.findMedian()).toBe(5.0);
  });

  it("two elements", () => {
    const mf = new MedianFinder();
    mf.addNum(1);
    expect(mf.findMedian()).toBe(1.0);
    mf.addNum(2);
    expect(mf.findMedian()).toBe(1.5);
  });

  it("negative numbers", () => {
    const mf = new MedianFinder();
    mf.addNum(-1);
    expect(mf.findMedian()).toBe(-1.0);
    mf.addNum(-2);
    expect(mf.findMedian()).toBe(-1.5);
    mf.addNum(-3);
    expect(mf.findMedian()).toBe(-2.0);
  });

  it("duplicates", () => {
    const mf = new MedianFinder();
    for (let i = 0; i < 4; i++) mf.addNum(1);
    expect(mf.findMedian()).toBe(1.0);
  });

  it("descending order", () => {
    const mf = new MedianFinder();
    const adds = [5, 4, 3, 2, 1];
    const medians = [5.0, 4.5, 4.0, 3.5, 3.0];
    for (let i = 0; i < adds.length; i++) {
      mf.addNum(adds[i]);
      expect(mf.findMedian()).toBe(medians[i]);
    }
  });

  it("large gap", () => {
    const mf = new MedianFinder();
    mf.addNum(0);
    expect(mf.findMedian()).toBe(0.0);
    mf.addNum(100);
    expect(mf.findMedian()).toBe(50.0);
  });
});
